File: elixir/grains/grains.exs
defmodule Grains do
  use Bitwise

  @doc """
  Calculate two to the power of the input minus one.
  """
  @spec square(pos_integer) :: pos_integer
  def square(number) do
    2 <<< (number - 2)
  end

  @doc """
  Adds square of each number from 1 to 64.
  """
  @spec total :: pos_integer
  def total do
    (1..64)
    |> Enum.map(&square/1)
    |> Enum.sum
  end
end

File: haskell/grains/src/Grains.hs
module Grains (square, total) where

import Data.Bits

square :: Integer -> Maybe Integer
square n
  | n < 1 || n > 64 = Nothing
  | otherwise       = Just (grains n)

grains :: Integer -> Integer
grains n = 2 ^ (n - 1)

total :: Integer
total = sum $ map grains [1..64]
